### PR TITLE
Fix rest type inference with tuples/closed arrays  

### PR DESCRIPTION
**Pull request by lasinicl**
## Purpose
$title

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/31142  Rest type of tuples/closed arrays should be created with the exact members instead of an open array
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/31296 Compiler crashes for a list-binding-pattern having a member list-binding-pattern with a rest-binding-pattern
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/31294 Tuple desructuring with a rest type doesn't work as expected

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 1]
```diff
@@ -2408,64 +2444,92 @@ public void visit(BLangTupleDestructure tupleDestructure) {
...
+        BLangExpression countExpr = ASTBuilderUtil.createLiteral(pos, symTable.intType,
+                (long) tupleVarRef.expressions.size());
+        boolean isIndexBasedAccessExpr = tupleExpr.getKind() == NodeKind.INDEX_BASED_ACCESS_EXPR;
+        if (restParamType.tag == TypeTags.TUPLE) {
+            BTupleType restTupleType = (BTupleType) restParamType;
+            BLangIndexBasedAccess modifiedTupleExpr = ASTBuilderUtil.createIndexAccessExpr(tupleExpr, countExpr);
+            modifiedTupleExpr.setBType(tupleExpr.getBType());
+            tupleExpr = modifiedTupleExpr;
+            isIndexBasedAccessExpr = true;
+            for (int index = 0; index < restTupleType.tupleTypes.size(); index++) {
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:2479]: Shall we move common logic used here and createRestFieldVarDefStmts() out

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 2]
```diff
@@ -2412,6 +2417,18 @@ private void assignTypesToMemberPatterns(BLangMatchPattern matchPattern, BType p
                     if (listMatchPattern.restMatchPattern == null) {
                         return;
                     }
+                    List<BType> memTypes = new ArrayList<>();
+                    if (arrayType.state == BArrayState.CLOSED) {
+                        for (int i = listMatchPattern.matchPatterns.size(); i < arrayType.size; i++) {
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:2422]: Try using Collections.nCopies() instead

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 3]
```diff
@@ -2909,6 +2936,18 @@ private void assignTypesToMemberPatterns(BLangBindingPattern bindingPattern, BTy
                     if (listBindingPattern.restBindingPattern == null) {
                         return;
                     }
+                    List<BType> memTypes = new ArrayList<>();
+                    if (arrayType.state == BArrayState.CLOSED) {
+                        for (int i = listBindingPattern.bindingPatterns.size(); i < arrayType.size; i++) {
+                            memTypes.add(arrayType.eType);
+                        }
+                        BTupleType restTupleType = new BTupleType(memTypes);
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:2944]: Try to move out the common logic for creating tuple type. used same in  2425

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 4]
```diff
@@ -2932,16 +2971,27 @@ private void assignTypesToMemberPatterns(BLangBindingPattern bindingPattern, BTy
...
                 }
-                BType restType = ((BArrayType) listBindingPattern.restBindingPattern.getBType()).eType;
+                List<BType> remainingTypes = new ArrayList<>();
                 for (int i = bindingPatterns.size(); i < types.size(); i++) {
-                    restType = this.types.mergeTypes(restType, types.get(i));
+                    remainingTypes.add(types.get(i));
                 }
-                if (bindingPatternTupleType.restType != null) {
-                    restType = this.types.mergeTypes(restType, bindingPatternTupleType.restType);
+                if (!remainingTypes.isEmpty()) {
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:2978]: Common logic between match and binding assignTypesToMemberPatterns()

**Comment by rdhananjaya** [comment]
Tested, changes looks good

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 1 above)
Comment (lasinicl) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:2479]: The logic is a bit different for creating the `var defs` for a rest field in a tuple variable and in a tuple restructure. Therefore it's difficult to move the seemingly common logic given in those methods out.

**Review by rdhananjaya — APPROVED** [review]

**Review by pcnfernando — APPROVED** [review]
Tested. LGTM

**Pull request merged**